### DISCUSSION TITLE
Appwrite Functions G4

### DISCUSSION THREAD

**Discussion by Meldiron** [appwrite/appwrite#5016] [Ideas]
![Functions_Gen4](https://user-images.githubusercontent.com/19310830/213177028-11521294-2a62-4058-aa67-c86bebfe817a.png)

# `helloWorld()` 👋

## 🎓 Background

Appwrite Functions let you extend Appwrite's backend capabilities with your own backend code written in your favourite language. Functions can help you with scheduled task such as cleanup, event-triggered tasks such as invoice generation, or http-triggered tasks to implement your own Appwrite "endpoints".

Functions are currently fast and powerful, but there are some downsides and limitations. To address them, we plan to introduce new generation of functions 🔥 This discussion describes our suggested changes. Keep in mind it's a discussion and nothing is set in stone.

## 💎 Backwards Compatibility

All existing functions will be backwards compatible, and you won't need to re-write function to a new generation.

All new functions made after upgrading Appwrite will need to be written in a new syntax.

### Hello world

**Current Generation:**

```js
module.exports = async (req, res) => {
  res.json({
    msg: 'Hello 👋',
  });
};
```

**New Generation:**

```js
module.exports = async (context) => {
  return context.res.json({
    msg: 'Hello 👋',
  });
};
```

> On first look, almost everything stays the same. Request and response objects are now under `context`, and you have to `return` when sending response 😇

In some languages you can even use [Destructuring assignment](https://www.javascripttutorial.net/es6/javascript-object-destructuring/), to only extract needed parts of `context`:

```js
module.exports = async ({ req, res }) => {
  return res.json({
    msg: 'Hello 👋',
  });
};
```

### Working with payload

**Current Generation:**

```js
module.exports = async (req, res) => {
  const name = JSON.parse(req.payload ?? '{}').name;
  res.send(`Your name is ${name}`);
};
```

**New Generation:**

```js
module.exports = async (context) => {
  const name = context.req.body.name;
  return context.res.send(`Your name is ${name}`);
};
```

> `payload` is now `body` so users from other backgrounds will understand the terminology automatically. Also notice that JSON bodies are automatically parsed for you 🎉

### Variables

```js
module.exports = async (req, res) => {
  const maxUsers = req.variables.MAX_USERS;
  res.send(`Your variable is ${maxUsers}`);
};
```

**New Generation:**

```js
module.exports = async (context) => {
  const maxUsers = process.env.MAX_USERS;
  return context.res.send(`Your variable is ${maxUsers}`);
};
```

> HTTP standard does not know variables. It was something introduced by Appwrite. We don't want developers to learn new concepts. We want to meet developers where they are, and so we removed variables form the request object. New generation will provide all variables as system environment variables. This is a common pattern seen in other frameworks, for instance [Next.JS](https://nextjs.org/) 🤝

### Logging

```js
module.exports = async (req, res) => {
  console.log('Function started');
  console.error('Whoops, something is wrong.');

  res.send('OK');
};
```

**New Generation:**

```js
module.exports = async (context) => {
  context.log('Function started');
  context.error('Whoops, something is wrong.');

  return context.res.send('OK');
};
```

> We introduced two logging functions under `context` instead of hooking into language native logging methods. This change allows more reliable per-execution logging that Appwrite community liked since it got introduced.

## 🔮 Future possibilities

New generation will introduce a custom URL for functions. For example, you could have URL `my-function.my-project.appwrite.myapp.com`. You can send request of any kind to this endpoint and your function will be triggered. This can be helpful when trying to use Appwrite without SDKs, or when using Appwrite function as webhook for payment gateway integration.

When using this function URLs, some more options become possible.

### Request headers

```js
module.exports = async (context) => {
  const stripeKey = context.request.headers['stripe-signature'];
  return context.res.send('Stripe webhook recieved');
};
```

### More request information

```js
module.exports = async (context) => {
  const { url, method, body } = context.req;
  return context.res.json({ url, method, body });
};
```

### HTML response

```js
const fs = require('fs');

// I am doing cool trick here, only reading file once during cold start. It's kept in memory for quick response to each execution
const indexFile = fs.readFileSync(__dirname + '/index.html').toString();

module.exports = async (context) => {
  return context.res.html(indexFile);
};
```

### File response

```js
module.exports = async (context) => {
  return context.res.file('kitty.png', 200, { 'content-type': 'image/png' });
};
```

### Redirect

```js
module.exports = async (context) => {
  // Pseudo-code
  const order = stripe.createNewOrder(products);
  const paymentUrl = order.publicUrl;
  
  // Real code
  return context.res.redirect(paymentUrl);
};
```

---

**Let us know what you think! 🔥**

We are satisfied with our implementation but we believe transparency is a key to success. We are opening this discussion to public and would love to hear what you love and what you hate about the new generation. Let's chat 😇

**Comment by monzim** [comment]
Are we going to see golang support for functions?

**Reply by stnguyen90** [reply]
This is almost ready! The [PR](https://github.com/open-runtimes/open-runtimes/pull/111) is just pending merge and a quick update in Appwrite!

**Comment by stnguyen90** [comment]
> Also notice that JSON bodies are automatically parsed for you 

@Meldiron, how does this work for Appwrite considering data is sent as a string? What if the developer doesn't want to pass a JSON string?

**Reply by tannermeade** [reply]
@Meldiron  @stnguyen90  If Appwrite parses the JSON for you Stripe webhook is impossible to implement because the security validation requires the function to have access to the raw unparsed text of the request.

**Reply by tannermeade** [reply]
Looks like they may have added an extra param so the function can have access to the raw body: https://github.com/open-runtimes/open-runtimes/blob/main/runtimes/node-16.0/src/server.js#L26

**Comment by stnguyen90** [comment]
> New generation will provide all variables as system environment variables. 

I think it might be worth mentioning some variables like `APPWRITE_FUNCTION_JWT` and `APPWRITE_FUNCTION_USER_ID` will end up being in the request headers.

**Comment by zlmr** [comment]
Would it be possible to add option to upload prebuilt binaries instead of code? So basically something that was possible in versions prior to 0.13.
That would reduce server resources by a lot since building takes a lot of CPU/RAM for some runtimes like Java/Kotlin.

Would also benefit you for Appwrite Cloud - I am using 4cores/8GB server and it is already [struggling](https://github.com/appwrite/appwrite/issues/5093) when I have to deploy 25 functions.

Not sure if the question is out of scope but since you are refactoring functions maybe it is worth giving it a look.

**Comment by rafagazani** [comment]
@Meldiron Do you have a release date?

**Reply by tannermeade** [reply]
I don't work at Appwrite, but here is the progress in github:

Functions V4 requires changes to the runtime environments (openruntimes). I believe this is the PR for the runtimes that'll be used in functions v4: https://github.com/open-runtimes/open-runtimes/pull/136 (merged)

I think this will be the main PR to get functions v4 done/merged into main: https://github.com/appwrite/docs/pull/398
- it's a draft PR right now, but looks like good progress.

**Comment by lucasctd** [comment]
`payload is now body so users from other backgrounds will understand the terminology automatically. Also notice that JSON bodies are automatically parsed for you 🎉`
Any plans to parse the response too?

That's what we have now:

```json
{
    "$id": "64165dbc616d0cxxxx95c",
    "$createdAt": "2023-03-19T00:56:28.410+00:00",
    "$updatedAt": "2023-03-19T00:56:43.844+00:00",
    "$permissions": [],
    "functionId": "functionId",
    "trigger": "http",
    "status": "completed",
    "statusCode": 422,
    "response": "{\"message\":\"My message.\"}",
    "stdout": "",
    "stderr": "",
    "duration": 0.105232
}
```

I wish it could look like this:

```json
{
    "$id": "64165dbc616d0cxxxx95c",
    "$createdAt": "2023-03-19T00:56:28.410+00:00",
    "$updatedAt": "2023-03-19T00:56:43.844+00:00",
    "$permissions": [],
    "functionId": "functionId",
    "trigger": "http",
    "status": "completed",
    "statusCode": 422,
    "response": {"message": "My Message"}
    "stdout": "",
    "stderr": "",
    "duration": 0.105232
}
```

**Comment by wilhantian** [comment]
Hi, I think customizing URLs is an urgent feature. When will it be released?

**Comment by Whitehat645** [comment]
when can we get this feature as this is a basic need, as we need to get web requests outside appwrite like payment gateways, what can we do without that guys, should we launch a ecom App for free.

**Comment by byawitz** [comment]
For anyone who needs to have URL access to your functions, you can try:

Function Proxy - https://github.com/stnguyen90/appwrite-function-proxy
Funcover -  https://github.com/BoolCode/appwrite-funcover

Funcover Docker container will provide you with anything you need (including headers) to use your Appwrite function as an input webhook.

**Comment by moshOntong-IT** [comment]
is this new generation can support implementing discord bot?

**Reply by byawitz** [reply]
Do you mean it will be possible to develop a Discord bot using the future Appwrite functions version?

**Reply by moshOntong-IT** [reply]
yes

**Reply by byawitz** [reply]
Then yes, the next version should enable developing any incoming webhooks using functions.

**Comment by Meldiron** [comment]
Thanks everyone for your feedback, generation 4 of Appwrite Functions was released in Appwrite 1.4 🎉
